### ISSUE TITLE
Gradle 7 incompatibility due to implicit dependency.

### ISSUE THREAD

**Issue by ZacSweers** [Kotlin/dokka#1833]
**Describe the bug**
Gradle 7.0 reports this issue for multiple dokka tasks

```
> Task :moshi-metadata-reflect:javadocsJar
Execution optimizations have been disabled for task ':moshi-metadata-reflect:javadocsJar' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/home/runner/work/MoshiX/MoshiX/docs/0.x'. Reason: Task ':moshi-metadata-reflect:javadocsJar' uses this output of task ':moshi-adapters:dokkaHtml' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to docs.gradle.org/7.0/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

**Expected behaviour**

Dokka should not have gradle issues

**Screenshots**

N/A

**To Reproduce**
Run Dokka with Gradle 7.0

Here's an example: https://github.com/ZacSweers/MoshiX/runs/2310971282

**Dokka configuration**
Configuration of dokka used to reproduce the bug

```kotlin
    tasks.named<DokkaTask>("dokkaHtml") {
      outputDirectory.set(rootProject.rootDir.resolve("docs/0.x"))
      dokkaSourceSets.configureEach {
        skipDeprecated.set(true)
        externalDocumentationLink {
          url.set(URL("https://square.github.io/okio/2.x/okio/"))
        }
        externalDocumentationLink {
          url.set(URL("https://square.github.io/moshi/1.x/moshi/"))
        }
      }
    }
```

**Installation**
- Operating system: N/A
- Build tool: Gradle v7
- Dokka version: 1.4.30

**Additional context**
None

**Comment by kamildoleglo** [comment]
Haven't got time to check that yet, but are you sure that this is not a problem on your side?
The fragment
> 'moshi-metadata-reflect:javadocsJar' uses this output of task ':moshi-adapters:dokkaHtml' without declaring an explicit or implicit dependency.

indicates that you have a task called `javadocsJar` that has a definition like (I'm guessing):

```
task javadocsJar { 
    from(dokkaHtml.outputDir.file)
}
```
while it should have a `dependsOn` also: 
```
task javadocsJar { 
    dependsOn(dokkaHtml)
    from(dokkaHtml.outputDir.file)
}
```

**Comment by ZacSweers** [comment]
The only thing that could be setting them is this plugin, which does set explicit dependencies: https://github.com/vanniktech/gradle-maven-publish-plugin/blob/0.14.2/src/main/kotlin/com/vanniktech/maven/publish/tasks/JavadocsJar.kt

**Comment by ZacSweers** [comment]
This looks like a wider issue with Dokka reading outputs from tasks in a generic way. Here is the same issue but using generated code from another plugin ([Wire](https://github.com/square/wire))

```
  - Gradle detected a problem with the following location: '/home/runner/work/slack-gradle-plugin/slack-gradle-plugin/slack-plugin/build/generated/source/wire'. Reason: Task ':slack-plugin:dokkaHtml' uses this output of task ':slack-plugin:generateMainProtos' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to docs.gradle.org/7.0/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

**Comment by Fachher** [comment]
Hi all.
Is there some flag to break the build if gradle detects a problem?

Thank you in advance

**Comment by chrisr3** [comment]
Hi, I have just tried upgrde to Gradle 7.1.1 but our Dokka 0.10.1 plugin fails with:
```
* What went wrong:
Some problems were found with the configuration of task ':dokka' (type 'DokkaTask').
  - In plugin 'org.jetbrains.dokka' type 'org.jetbrains.dokka.gradle.DokkaTask' property 'configuration.collectKotlinTasks' is missing an input or output annotation.
    
    Reason: A property without annotation isn't considered during up-to-date checking.
    
    Possible solutions:
      1. Add an input or output annotation.
      2. Mark it as @Internal.
    
    Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#missing_annotation for more details about this problem.
  - In plugin 'org.jetbrains.dokka' type 'org.jetbrains.dokka.gradle.DokkaTask' property 'configuration.name' is missing an input or output annotation.
    
    Reason: A property without annotation isn't considered during up-to-date checking.
    
    Possible solutions:
      1. Add an input or output annotation.
      2. Mark it as @Internal.
    
    Please refer to https://docs.gradle.org/7.1.1/userguide/validation_problems.html#missing_annotation for more details about this problem.
  - In plugin 'org.jetbrains.dokka' type 'org.jetbrains.dokka.gradle.DokkaTask' property 'dokkaRuntime' is missing an input or output annotation.
    
    Reason: A property without annotation isn't considered during up-to-date checking.
    
    Possible solutions:
      1. Add an input or output annotation.
      2. Mark it as @Internal.
```
Unfortunately, upgrading to Dokka 1.4.30 breaks due to OOM, and even **2GB** of Metaspace isn't enough to avoid that. Seeing as the latest Dokka plugin has an apparently _unfixable_ memory leak, can someone release a 0.10.2 version that is compatible with Gradle 7.1 please?

**Comment by JamshedAlamQaderi** [comment]
> This looks like a wider issue with Dokka reading outputs from tasks in a generic way. Here is the same issue but using generated code from another plugin ([Wire](https://github.com/square/wire))
> 
> ```
>   - Gradle detected a problem with the following location: '/home/runner/work/slack-gradle-plugin/slack-gradle-plugin/slack-plugin/build/generated/source/wire'. Reason: Task ':slack-plugin:dokkaHtml' uses this output of task ':slack-plugin:generateMainProtos' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to docs.gradle.org/7.0/userguide/validation_problems.html#implicit_dependency for more details about this problem.
> ```

Hi @ZacSweers ,
How did you solved this issue?

**Comment by CyxouD** [comment]
I got this error when I replaced `dependsOn` with `.mustRunAfter`.  It was needed for other task to fix error "Cannot change dependencies of dependency configuration after it has been resolved." after downgrading to Gradle plugin 7.2.2

**Comment by whyoleg** [comment]
Should not be an issue in [Dokka 2.0.0](https://github.com/Kotlin/dokka/releases/tag/v2.0.0) with [Dokka Gradle plugin v2﻿](https://kotlinlang.org/docs/dokka-migration.html)

**Comment by dkhalanskyjb** [comment]
Reproduces in `kotlinx.coroutines` if we remove the workaround, even though Dokka 2.0.0 is used with
```gradle
org.jetbrains.dokka.experimental.gradle.pluginMode=V2Enabled
org.jetbrains.dokka.experimental.gradle.pluginMode.nowarn=true
```
In `kotlinx-coroutines-core/build.gradle.kts`, we have
```kotlin
tasks.withType<DokkaBaseTask>() {
    dependsOn(jvmJar)
}
```
Removing it makes the build fail with
```
> Task :kotlinx-coroutines-core:dokkaGenerateModuleHtml FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':kotlinx-coroutines-core:dokkaGenerateModuleHtml' (type 'DokkaGenerateModuleTask').
  - Gradle detected a problem with the following location: '/home/jb/IdeaProjects/kotlinx.coroutines/kotlinx-coroutines-core/build/libs/kotlinx-coroutines-core-jvm-1.10.2-SNAPSHOT.jar'.

    Reason: Task ':kotlinx-coroutines-core:dokkaGenerateModuleHtml' uses this output of task ':kotlinx-coroutines-core:jvmJar' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':kotlinx-coroutines-core:jvmJar' as an input of ':kotlinx-coroutines-core:dokkaGenerateModuleHtml'.
      2. Declare an explicit dependency on ':kotlinx-coroutines-core:jvmJar' from ':kotlinx-coroutines-core:dokkaGenerateModuleHtml' using Task#dependsOn.
      3. Declare an explicit dependency on ':kotlinx-coroutines-core:jvmJar' from ':kotlinx-coroutines-core:dokkaGenerateModuleHtml' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.13/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

**Comment by whyoleg** [comment]
@dkhalanskyjb, most likely this is a new issue related to the coroutine setup. AFAIK, last time @adam-enko was investigating this issue, it was not reproducible when running the task explicitly, but only when running the ' root `:dokkaGenerate` task, meaning that there is some cross-project configuration happening, making diagnosing this harder...

**Comment by adam-enko** [comment]
@dkhalanskyjb I checked out the branch https://github.com/Kotlin/kotlinx.coroutines/tree/dgp-2.0 to investigate.

I commented out the `dokka-conventions.gradle.kts` from buildscripts, because I expect that DGP should work well without any configuration. This caused some errors in the buildscripts, but after fixing them, and manually configuring the Dokka aggregation, I ran into the expected issue [KT-71784 Illegal use of flexible type deserializer](https://youtrack.jetbrains.com/issue/KT-71784/DGP-v2-dokkaGenerateModuleHtml-for-coroutines-core-fails-with-Illegal-use-of-flexible-type-deserializer). KT-71784 is fixed in Dokka 2.1.0, so I updated the version in coroutines' `gradle.properties`. Re-running `./gradlew :dokkaGenerate` was successful.

Therefore, the cause must be in `dokka-conventions.gradle.kts`. I removed the workaround for KT-71784 (suppressing the `commonMain` DokkaSourceSet), since it's fixed in the newer Dokka version. This resolved the issue.

https://github.com/Kotlin/kotlinx.coroutines/blob/d5cea5c7c14b53e4baeadb83c4c22d3ed35fa449/buildSrc/src/main/kotlin/dokka-conventions.gradle.kts#L62-L65

I'm not quite sure why suppressing `commonMain` causes this issue. It is intended that the dependencies of suppressed sources are not resolved, otherwise DGP will trigger unnecessary work. However, it could be considered a DGP issue (suppressing source sets should 'just work'). Since here it was caused by a workaround for a since fixed issue, and there might be another cause (the build config of coroutines does not follow best practices), I don't think we should take further action on the Dokka side.

In summary:

1. Bump the Dokka version to 2.1.0.
2. Remove the workaround for KT-71784.

**Comment by dkhalanskyjb** [comment]
@adam-enko, thanks! That branch is very outdated, though. There isn't such a workaround anymore, and `dokka-conventions.gradle.kts` looks very different. With today's PR to upgrade `kotlinx.coroutines` to 2.2.0-Beta (https://github.com/Kotlin/kotlinx.coroutines/pull/4617), I still encounter the problem if I remove the `dependsOn` workaround (https://github.com/Kotlin/dokka/issues/1833#issuecomment-3823887097).

**Comment by adam-enko** [comment]
@dkhalanskyjb Thanks for the update. I commented out some more of the non-idiomatic build config and it appears to resolve the issue, which indicates there's something else amiss in the build config. Perhaps another plugin is modifying the jvmJar task.

https://github.com/Kotlin/kotlinx.coroutines/commit/de9343ab868c0bb5afd9d7962e7f1edac423ac44

To investigate further we'd need a smaller reproducer. The coroutines project uses a lot of discouraged practices in its build config which makes it hard to untangle.

**Comment by dkhalanskyjb** [comment]
> Perhaps another plugin is modifying the jvmJar task

Yes, it does: https://github.com/Kotlin/kotlinx.coroutines/blob/b11abdf01d4d5db85247ab365abc72efc7b95062/buildSrc/src/main/kotlin/version-file-conventions.gradle.kts and (not a plugin) https://github.com/Kotlin/kotlinx.coroutines/blob/b11abdf01d4d5db85247ab365abc72efc7b95062/kotlinx-coroutines-core/build.gradle.kts#L154

I'm unfortunately not proficient enough with Gradle to understand the problem or how it can be fixed, though. Not that any of this causes us any issues!
